pkg/util/active_groups_test.go

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
+	"golang.org/x/exp/slices"
+)
+
+const maxGroupsPerUser = 100
+
+func TestActiveGroups(t *testing.T) {
+	ag := NewActiveGroups(maxGroupsPerUser)
+
+	ag.UpdateGroupTimestampForUser("user1", "group1", time.Unix(0, 10))
+	ag.UpdateGroupTimestampForUser("user1", "group2", time.Unix(0, 15))
+	ag.UpdateGroupTimestampForUser("user1", "group3", time.Unix(0, 20))
+
+	ag.UpdateGroupTimestampForUser("user2", "group4", time.Unix(0, 10))
+	ag.UpdateGroupTimestampForUser("user2", "group5", time.Unix(0, 15))
+	ag.UpdateGroupTimestampForUser("user2", "group6", time.Unix(0, 20))
+
+	require.Nil(t, ag.PurgeInactiveGroupsForUser("user1", 5))
+	inactiveGroupsForUser1 := ag.PurgeInactiveGroupsForUser("user1", 16)
+	require.Equal(t, 2, len(inactiveGroupsForUser1))
+	require.True(t, slices.Contains(inactiveGroupsForUser1, "group1"))
+	require.True(t, slices.Contains(inactiveGroupsForUser1, "group2"))
+	require.False(t, slices.Contains(inactiveGroupsForUser1, "group3"))
+
+	require.Nil(t, ag.PurgeInactiveGroupsForUser("user2", 5))
+	inactiveGroupsForUser2 := ag.PurgeInactiveGroupsForUser("user2", 11)
+	require.Equal(t, 1, len(inactiveGroupsForUser2))
+	require.True(t, slices.Contains(inactiveGroupsForUser2, "group4"))
+	require.False(t, slices.Contains(inactiveGroupsForUser2, "group5"))
+	require.False(t, slices.Contains(inactiveGroupsForUser2, "group6"))
+
+	ag.UpdateGroupTimestampForUser("user1", "group1", time.Unix(0, 25))
+	inactiveGroupsForUser1 = ag.PurgeInactiveGroupsForUser("user1", 21)
+	require.Equal(t, 1, len(inactiveGroupsForUser1))
+	require.Equal(t, []string{"group3"}, inactiveGroupsForUser1)
+}
+
+func TestActiveGroupsConcurrentUpdateAndPurge(t *testing.T) {
+	numGroups := 10
+	ag := NewActiveGroups(maxGroupsPerUser)
+	done := sync.WaitGroup{}
+	stop := atomic.NewBool(false)
+	latestTS := atomic.NewInt64(0)
+
+	for i := 0; i < numGroups; i++ {
+		done.Add(1)
+		go func() {
+			defer done.Done()
+
+			for !stop.Load() {
+				ts := latestTS.Inc()
+				ag.UpdateGroupTimestampForUser("user1", fmt.Sprintf("%d", ts), time.Unix(0, ts))
+				time.Sleep(1 * time.Millisecond)
+			}
+		}()
+	}
+
+	previousLatest := int64(0)
+	for i := 0; i < numGroups; i++ {
+		time.Sleep(100 * time.Millisecond)
+
+		latest := latestTS.Load()
+		require.True(t, latest > previousLatest)
```

Comment:
If this fails it's unclear whether it's a broken test assumption or a broken test. Add a failing message please.